### DISCUSSION TITLE
Scalability [TLS] - Caching and better indexing Destination Rules for TLS computing

### DISCUSSION THREAD

**Discussion by xeviknal** [kiali/kiali#4509] [Roadmap]
In order to compute with 100% accuracy whether a namespace, a mesh, a service has mTLS enabled, you need to parse all the DestinationRules visible at the subject scope. Plus, DestinationRules are visible in the whole service mesh by default.

Having a high volume of Destination Rule makes this computation pretty costly. And if you add up that kiali needs to trigger as many k8s API calls as namespaces kiali has available, it even increases the response time.

But using autoMTLS you don't need to define DestinationRules, right?
AutoMTLS removes the need of adding DestinationRules for enabling mTLS. However, users need to add Destination Rules to disable mTLS or use a different TLS type. Hence, in order to be 100% accurate while computing the mTLS status, kiali needs to visit the related DRs to the host/namespace/mesh.

Therefore, I propose this discussion to talk about how this should be approached.

I have a proposal to share.

1) Use a different data structure to store the Destination Rules in kiali for reducing time in the lookup.
Something like a map `map[{host,tlsMode}][]*DestinationRule` would definitely speed up when looking up for DestinationRules pointing to `host: *.bookinfo.svc.cluster.local` or to `*.local`. There wouldn't be the need to visit all the DestinationRules to get the information.

2) Keep this structure in-memory. There are different features using it: tls at mesh level, tls at namespace level, computing all the validations around mtls misconfigurations. Also a new feature could be useful to compute as a consequence: tls at service level.
Kiali could follow 2 expiration approaches for this cache: 
2.1) re-compute from scratch when a DestinationRule is added/modified/removed
2.2) update the structure via informers for DestinationRules.

3) Cache the results: kiali could keep a cache for the tls endpoints (ns and mesh-level). This cache could be expired via 2.1/2.2.

Tackle this: https://github.com/kiali/kiali/issues/4224
Expanding this comment: https://github.com/kiali/kiali/discussions/4080#discussioncomment-1063853
This [dummy script](https://github.com/xeviknal/istio-lab/blob/master/kiali-4224/dest-rule-adder.sh) slows down the tls and validations endpoint.

**Comment by xeviknal** [comment]
There is another approach that I haven't explored that much: using the `/debug/endpointz` endpoint or `/debug/edsz?proxyID=reviews-v2-7bf8c9648f-vvgrr.bookinfo` one. It is shown the `tlsMode` for each endpoint. Therefore, doing some filters we could definitely whether a ns or a mesh is using mtls.

![Screenshot of localhost_15014_debug_edsz_proxyID=reviews-v2-7bf8c9648f-vvgrr bookinfo](https://user-images.githubusercontent.com/613814/142479634-305a3bf5-2dcc-4d6c-8ac7-6af2689b4c61.png)

And just thinking out loud, we may even try to add a new debug endpoint into istiod pod to fetch this information. Istiod (pilot) has the TLS information processed somewhere. If we could retrieve them and structure them in a single endpoint it would be a no brainer for kiali to fetch and render.

**Reply by jshaughn** [reply]
If we can avoid the namespace-scoped k8s API calls it would be a win, right (or if we can use cluster perms)?   We already use the `/debug` endpoint for access to the proxy sync and maybe registry data, so perhaps we should explore this approach?  Although, we also don't want to be too heavy on pilot, I guess.   Regardless of approach I think I like the idea of caching the information.

**Reply by lucasponce** [reply]
Note that the Istio Registry endpoints are introduced in the https://github.com/kiali/kiali/pull/4479 as they will be needed for Service details.
We haven't yet to get deep into https://github.com/kiali/kiali/issues/4382 which aims to better use the Istio Registry info for validations.

It sounds promising, perhaps one gap I see is that it may be lost the case that a namespace is empty or with headless workloads, but it seems the way to go.

**Reply by xeviknal** [reply]
Yeah, it looks like it remove some work that is already done in pilot side.
However, using this info might imply a feature concept tweak. We would me moving from configuration status info to runtime information. We would move from saying: "the mesh only allows mTLS traffic (there are the expected configs in the admin namespace)" to "all the traffic detected in your mesh uses mtls (because all the envoys are config to send/receive mtls traffic)".

**Reply by lucasponce** [reply]
Yes, the configuration information is something we don't want to miss here.

I see this discussion more "actionable" as there is already an issue opened about it https://github.com/kiali/kiali/issues/4224, also all the work done in the Istio Registry will help.

Additionally we have a tls_perf_test.go in the code, that combined with the profiler, is helping to debug some aspects that can be improved [1].

And when the Istio configuration fetched from the registry is consolidated into the validations as part of the https://github.com/kiali/kiali/issues/4382 task this idea can be added as part of that process.

[1] 
![image](https://user-images.githubusercontent.com/1662329/142604276-86e21c9c-ea90-4b79-998c-439bedb13bf4.png)
